core/trino-main/src/main/java/io/trino/cost/PartialAggregationStatsRule.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.cost;
+
+import io.trino.Session;
+import io.trino.matching.Pattern;
+import io.trino.sql.planner.Symbol;
+import io.trino.sql.planner.TypeProvider;
+import io.trino.sql.planner.iterative.Lookup;
+import io.trino.sql.planner.plan.AggregationNode;
+import io.trino.sql.planner.plan.AggregationNode.Aggregation;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Optional;
+
+import static io.trino.sql.planner.plan.AggregationNode.Step.PARTIAL;
+import static io.trino.sql.planner.plan.Patterns.aggregation;
+
+public class PartialAggregationStatsRule
```

Comment:
What's the difference in adding a new rule vs handling the partial aggregation step explicitly in the existing AggregationStatsRule ?

Reply:
Partial aggregation need to be estimated before final aggregation to present symbol statistics that final aggregation is based on. With one step, you don't have possibility to estimate partial aggregations before final (you can, but this would require to do lookups for source nodes under partial aggregation node, this would be more complicated than separate rule).